File: src/main/scala/fr/xebia/streams/processing/DetectMotion.scala
package fr.xebia.streams.processing

import akka.stream.scaladsl.Flow
import fr.xebia.streams.transform.MediaConversion
import fr.xebia.streams.transform.MediaConversion.matToIplImage
import org.bytedeco.javacpp.opencv_core.{IplImage, Mat}

object DetectMotion {

  def apply(): Flow[Seq[Mat], Mat, _] = {
    Flow[Seq[Mat]]
      .collect { case l if l.size == 2 => (l.head, l.last) }
      .map {
        case (firstMat, secondMat) =>
          val first = matToIplImage(firstMat)
          val second = matToIplImage(secondMat)
          (first, second)
      }
      .map {
        case (first, second) =>
          import org.bytedeco.javacpp.opencv_core.{IplImage, cvAbsDiff}
          val firstSmoothed = smooth(first)
          val secondSmoothed = smooth(second)

          val diff: IplImage = second.clone()
          cvAbsDiff(firstSmoothed, secondSmoothed, diff)
          diff
      }
      .map { diff =>
        import org.bytedeco.javacpp.opencv_imgproc.{CV_THRESH_BINARY, cvThreshold}
        cvThreshold(diff, diff, 64, 255, CV_THRESH_BINARY)
        diff
      }
      .map(MediaConversion.iplImageToMat)
  }

  private def smooth(baseImage: IplImage): IplImage = {
    import org.bytedeco.javacpp.opencv_imgproc.{CV_GAUSSIAN, cvSmooth}
    val newImage: IplImage = baseImage.clone()
    cvSmooth(baseImage, newImage, CV_GAUSSIAN, 9, 9, 1, 1)
    newImage
  }


}

File: src/main/java/fr/xebia/streams/java/processing/DetectMotion.java
package fr.xebia.streams.java.processing;

import akka.NotUsed;
import akka.japi.Pair;
import akka.stream.javadsl.Flow;
import fr.xebia.streams.java.transform.MediaConversion;
import org.bytedeco.javacpp.opencv_core;
import org.bytedeco.javacpp.opencv_core.IplImage;

import java.util.List;

import static fr.xebia.streams.java.transform.MediaConversion.matToIplImage;
import static org.bytedeco.javacpp.opencv_core.cvAbsDiff;
import static org.bytedeco.javacpp.opencv_imgproc.*;

public class DetectMotion {

    public static Flow<List<opencv_core.Mat>, opencv_core.Mat, NotUsed> build() {
        return Flow.<List<opencv_core.Mat>>create()
                .filter(l -> l.size() == 2)
                .map(l -> new Pair<>(l.get(0), l.get(1)))
                .map(p -> new Pair<>(matToIplImage(p.first()), matToIplImage(p.second())))
                .map(p -> {
                    IplImage smoothFirst = smooth(p.first());
                    IplImage smoothSecond = smooth(p.second());
                    IplImage diff = smoothFirst.clone();
                    cvAbsDiff(smoothFirst, smoothSecond, diff);
                    return diff;
                })
                .map(diff -> {
                    cvThreshold(diff, diff, 64, 255, CV_THRESH_BINARY);
                    return diff;
                })
                .map(MediaConversion::iplImageToMat);
    }

    private static IplImage smooth(IplImage baseImage) {
        IplImage newImage = baseImage.clone();
        cvSmooth(baseImage, newImage, CV_GAUSSIAN, 9, 9, 1, 1);
        return newImage;
    }

}
